src/hotspot/cpu/aarch64/register_aarch64.hpp

```diff
@@ -124,45 +144,50 @@ CONSTANT_REGISTER_DECLARATION(Register, sp,  (33));
 // Used as a filler in instructions where a register field is unused.
 const Register dummy_reg = r31_sp;
 
+
 // Use FloatRegister as shortcut
 class FloatRegisterImpl;
 typedef FloatRegisterImpl* FloatRegister;
 
-inline FloatRegister as_FloatRegister(int encoding) {
-  return (FloatRegister)(intptr_t) encoding;
-}
+const FloatRegister as_FloatRegister(int encoding);
 
 // The implementation of floating point registers for the architecture
 class FloatRegisterImpl: public AbstractRegisterImpl {
- public:
+  const friend FloatRegister as_FloatRegister(int encoding);
+
+private:
+  static constexpr FloatRegister first();
+  static constexpr FloatRegister invalid() { return first() + number_of_declared_registers; }
+
+public:
   enum {
     number_of_registers = 32,
+    number_of_declared_registers = 32,
     max_slots_per_register = 8,
     save_slots_per_register = 2,
     slots_per_neon_register = 4,
     extra_save_slots_per_neon_register = slots_per_neon_register - save_slots_per_register
   };
 
-  // construction
-  inline friend FloatRegister as_FloatRegister(int encoding);
-
   VMReg as_VMReg();
 
   // derived registers, offsets, and addresses
-  FloatRegister successor() const                          { return as_FloatRegister((encoding() + 1) % 32); }
+  FloatRegister successor() const { return as_FloatRegister(encoding() + 1); }
 
   // accessors
-  int   encoding() const                          { assert(is_valid(), "invalid register"); return (intptr_t)this; }
-  int   encoding_nocheck() const                         { return (intptr_t)this; }
-  bool  is_valid() const                          { return 0 <= (intptr_t)this && (intptr_t)this < number_of_registers; }
+  bool is_valid() const           { return this < invalid(); }
+  bool has_byte_register() const  { return is_valid(); }
```

Comment:
Is `has_byte_register` needed here? I see it in the previous code for `Register` but not for `FloatRegister`.
```suggestion
```

Reply:
Done.